### PR TITLE
[Core] Run only one forceSizeUpdate per 16 milliseconds

### LINKED ISSUE #2039: ForceUpdateSize Delay issue
Hi Xamarin

https://github.com/xamarin/Xamarin.Forms/blob/master/Xamarin.Forms.Core/Cells/Cell.cs#L211-L218

We can see such comment "don't run more than once per 16 milliseconds", but this code doesn't ensure 16 milliseconds delay... It just requires to wait every thread, there is no thread sync..

Perhaps, it can be regarding to https://forums.xamarin.com/discussion/123437/forceupdatesize-slows-ui#latest

### PR DESCRIPTION
**Pull request by AndreiMisiukevich**
### Description of Change ###

fixes #2039 

Changed:
 - object Cell => OnForceUpdateSizeRequested

### Behavioral Changes ###
Run only one forceSizeUpdate per 16 milliseconds

### CODE REVIEW

File: Xamarin.Forms.Core/Cells/Cell.cs [hunk 1]
```diff
@@ -19,6 +19,9 @@ public abstract class Cell : Element, ICellController, IFlowDirectionController
 
 		bool _nextCallToForceUpdateSizeQueued;
 
+        object _forceUpdateSizeLocker = new object();
```
Comment (adrianknight89) [inline Xamarin.Forms.Core/Cells/Cell.cs]: Seems like indentation is out of place here and on other lines as well.

**Comment by AndreiMisiukevich** [comment]
@adrianknight89 sorry, it was my gap.. I had to uncheck "converting tabs to spaces"

**Comment by jassmith** [comment]
build

**Comment by jassmith** [comment]
I dont understand the point of this? Screen refresh rates can be greater than 60hz

File: Xamarin.Forms.Core/Cells/Cell.cs [hunk 2]
```diff
@@ -19,6 +19,9 @@ public abstract class Cell : Element, ICellController, IFlowDirectionController
 
 		bool _nextCallToForceUpdateSizeQueued;
 
+		object _forceUpdateSizeLocker = new object();
+		Task _forceUpdateSizeDelayTask = Task.CompletedTask;
```
Comment (jassmith) [inline Xamarin.Forms.Core/Cells/Cell.cs]: Compile error here:

Cells\Cell.cs(23,41): error CS0117: 'Task' does not contain a definition for 'CompletedTask' [C:\agent\_work\2\s\Xamarin.Forms.Core\Xamarin.Forms.Core.csproj]

**Comment by AndreiMisiukevich** [comment]
@jassmith I can't answer, nevertheless I noticed, that the implementation doesn't correspond to presented comment - "don't run more than once per 16 milliseconds"

look through initial code
https://github.com/xamarin/Xamarin.Forms/blob/master/Xamarin.Forms.Core/Cells/Cell.cs#L211-L218

So, I aligned code according to presented comment.

File: Xamarin.Forms.Core/Cells/Cell.cs (hunk 2 above)
Comment (AndreiMisiukevich) [reply Xamarin.Forms.Core/Cells/Cell.cs]: Fixed this error by replacing Task.CompletedTask by Task.FromResult(true)

**Comment by AndreiMisiukevich** [comment]
Now, I hope, it will be built successfully

**Comment by jassmith** [comment]
I apologize, I was skipping through these notifications very fast and failed to properly read the diff. I remember why this is rate limited now. Its SUPER expensive :P

**Comment by jassmith** [comment]
build

**Comment by rmarinho** [comment]
@AndreiMisiukevich unit test failed

**Comment by hartez** [comment]
@AndreiMisiukevich I do not understand why the original version would run more than once every 16 milliseconds. Are you worried about this being called from multiple threads?

**Comment by hartez** [comment]
@AndreiMisiukevich @jassmith I am guessing that most handlers of that event need to run on the main thread. Perhaps it would make more sense to marshal all the calls to OnForceUpdateSizeRequested onto the main thread in the first place. Something like:

```
public void ForceUpdateSize()
{
	if (_nextCallToForceUpdateSizeQueued)
		return;

	if ((Parent as ListView)?.HasUnevenRows == true)
	{
		Device.BeginInvokeOnMainThread(() => {
			_nextCallToForceUpdateSizeQueued = true;
			OnForceUpdateSizeRequested();
		});
	}
}
```

**Comment by AndreiMisiukevich** [comment]
Okay, i see.
But I think it would be better if we wait only necessary time (without default 16 milliseconds waiting every call)

I added new commit. see it above

**Comment by jassmith** [comment]
Thats not a bad point. Restarting your builds again. You probably should just null check that Task instead of Task.Delay(0) it, that is just waste.

**Comment by AndreiMisiukevich** [comment]
so @jassmith, what do you think about Device.BeginInvokeInMainThread(OnForceUpdateSizeRequested) as @hartz said?

**Comment by jassmith** [comment]
I dont want us to auto-marshal everything no. Supporting out of UI thread calls in just this one area will cause us more confusion than its worth.

**Comment by AndreiMisiukevich** [comment]
Colleagues, I think these changes aren't very important, are they?
Should I close/discard this PR?

**Comment by rmarinho** [comment]
Does it show any visual change' like is the list faster , smoother ?

**Comment by rmarinho** [comment]
We will not be taking this at this time.

**Comment by karlingen** [comment]
@AndreiMisiukevich @rmarinho Why is this closed? This is still an issue. See https://github.com/xamarin/Xamarin.Forms/issues/2735

**Pull request closed without merging**